aries_cloudagent/multitenant/route_manager.py

```diff
@@ -0,0 +1,105 @@
+"""Multitenancy route manager."""
+
+
+import logging
+from typing import List, Optional, Tuple
+
+from ..core.profile import Profile
+from ..messaging.responder import BaseResponder
+from ..protocols.coordinate_mediation.v1_0.manager import MediationManager
+from ..protocols.coordinate_mediation.v1_0.models.mediation_record import (
+    MediationRecord,
+)
+from ..protocols.coordinate_mediation.v1_0.route_manager import RouteManager
+from ..protocols.routing.v1_0.manager import RoutingManager
+from ..protocols.routing.v1_0.models.route_record import RouteRecord
+from ..storage.error import StorageNotFoundError
+
+
+LOGGER = logging.getLogger(__name__)
+
+
+class MultitenantRouteManager(RouteManager):
+    """Multitenancy route manager."""
+
+    def __init__(
+        self,
+        root_profile: Profile,
+    ):
+        """Initialize multitenant route manager."""
+        self.root_profile = root_profile
+
+    async def get_base_wallet_mediator(self) -> Optional[MediationRecord]:
+        """Get base wallet's default mediator."""
+        return await MediationManager(self.root_profile).get_default_mediator()
+
+    async def _route_for_key(
+        self,
+        profile: Profile,
+        recipient_key: str,
+        mediation_record: Optional[MediationRecord] = None,
+        *,
+        skip_if_exists: bool = False,
+        replace_key: Optional[str] = None,
+    ):
+        wallet_id = profile.settings["wallet.id"]
+        LOGGER.info(
+            f"Add route record for recipient {recipient_key} to wallet {wallet_id}"
+        )
+        routing_mgr = RoutingManager(self.root_profile)
+        mediation_mgr = MediationManager(self.root_profile)
+        # If base wallet had mediator, only notify that mediator.
+        # Else, if subwallet has mediator, notify that mediator.
+        base_mediation_record = await self.get_base_wallet_mediator()
+        mediation_record = base_mediation_record or mediation_record
+
+        if skip_if_exists:
+            try:
+                async with self.root_profile.session() as session:
+                    await RouteRecord.retrieve_by_recipient_key(session, recipient_key)
+
+                # If no error is thrown, it means there is already a record
+                return None
+            except (StorageNotFoundError):
+                pass
+
+        await routing_mgr.create_route_record(
+            recipient_key=recipient_key, internal_wallet_id=wallet_id
+        )
+
+        # External mediation
+        keylist_updates = None
+        if mediation_record:
```

Comment:
It looks like this may be performing key updates with the base wallet mediator, which the old code explicitly avoids. Or does the mediation manager skip it when not necessary?

Reply:
Old add_key method in multitenant manager:
```python
        mediation_record = await mediation_mgr.get_default_mediator()

        if skip_if_exists:
            try:
                async with self._profile.session() as session:
                    await RouteRecord.retrieve_by_recipient_key(session, recipient_key)

                # If no error is thrown, it means there is already a record
                return
            except (StorageNotFoundError):
                pass

        await routing_mgr.create_route_record(
            recipient_key=recipient_key, internal_wallet_id=wallet_id
        )

        # External mediation
        if mediation_record:
            keylist_updates = await mediation_mgr.add_key(recipient_key)

            responder = self._profile.inject(BaseResponder)
            await responder.send(
                keylist_updates, connection_id=mediation_record.connection_id
            )
```

So this old method looks very similar to what's currently implemented. However, the old code only dealt with sending keylist updates to the base wallet mediator when needed. The old code paths through the different connection forming protocols would trigger keylist updates to a sub wallet mediator independent of the multi-tenant `add_key` method. This actually resulted in both the sub wallet and base wallet mediator getting keylist updates when in reality only the base wallet mediator should have been receiving keylist updates.

This new method will always send keylist updates to a base wallet mediator, if present. It will send a keylist update to a sub wallet mediator iff no base wallet mediator is set. And, of course, this results in the correct behavior because all of the keylist update sending behavior has been removed from the connection forming protocols and delegated to the route managers.

Does that clear things up?

Reply:
I'm thinking of this logic in the connections manager: https://github.com/hyperledger/aries-cloudagent-python/blob/f0b27196614a3d196af82ee6927d6c9bfab58a2f/aries_cloudagent/protocols/connections/v1_0/manager.py#L241

Reply:
Ah, okay, I must have missed that line. So the old code did send keylist updates to only the base wallet mediator, at least in that instance. The `keylist_updates` getting nulled out at the chunk you linked to ensures keylist updates aren't sent from the connection protocol methods because they would instead be sent by the multi-tenant manager `add_key` method.

I'm pretty confident the same logic is captured correctly in the multi-tenant route manager -- it's just all in one place now